partiql-parser/src/result.rs

```diff
@@ -4,33 +4,140 @@
 
 use pest::error::{ErrorVariant, LineColLocation};
 use std::fmt;
+use std::fmt::Formatter;
 use thiserror::Error;
 
-/// Position in the source for an error.
+/// A line and column location.
+///
+/// This value is one-based, as that is how most people think of lines and columns.
+///
+/// ## Example
+/// ```
+/// # use partiql_parser::prelude::*;
+/// println!("Beginning of a document: {}", LineAndColumn::at(1, 1));
+/// ```
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct LineAndColumn(pub usize, pub usize);
+
+impl LineAndColumn {
+    /// Constructs a [`LineAndColumn`].
+    #[inline]
+    pub fn at(line: usize, column: usize) -> Self {
+        Self(line, column)
+    }
+
+    /// Returns a [`LineAndColumn`] that moves to the given position relative
+    /// to this one as a sort of "origin."
+    ///
+    /// Note that this positioning is 1-based, so moving `(1, 1)` to `(1, 1)` is a no-op.
+    ///
+    /// ## Examples
+    /// ```
+    /// # use partiql_parser::prelude::*;
+    /// assert_eq!(
+    ///     LineAndColumn::at(1, 1),
+    ///     LineAndColumn::at(1, 1).move_relative(LineAndColumn::at(1, 1))
+    /// );
+    /// ```
+    ///
+    /// ```
+    /// # use partiql_parser::prelude::*;
+    /// assert_eq!(
+    ///     LineAndColumn::at(1, 2),
+    ///     LineAndColumn::at(1, 1).move_relative(LineAndColumn::at(1, 2))
+    /// );
+    /// ```
+    ///
+    /// ```
+    /// # use partiql_parser::prelude::*;
+    /// assert_eq!(
+    ///     LineAndColumn::at(5, 10),
+    ///     LineAndColumn::at(5, 7).move_relative(LineAndColumn::at(1, 4))
+    /// );
+    /// ```
+    ///
+    /// ```
+    /// # use partiql_parser::prelude::*;
+    /// assert_eq!(
+    ///     LineAndColumn::at(21, 2),
+    ///     LineAndColumn::at(2, 15).move_relative(LineAndColumn::at(20, 2))
+    /// );
+    /// ```
+    pub fn move_relative(self, location: LineAndColumn) -> Self {
```

Comment:
This API is needed so we can use the PEG "continuation" style.  Call in, remember where we left off, and re-translate the parsers line/column positioning because every time we call into the PEG, it thinks we're starting at `(1, 1)`.